src/script/src/python/pyo3/utils.rs

```diff
@@ -0,0 +1,259 @@
+            list.to_object(py)
+        }
+    })
+}
+
+macro_rules! to_con_type {
+    ($dtype:ident,$obj:ident, $($cty:ident => $rty:ty),*$(,)?) => {
+        match $dtype {
+            $(
+                ConcreteDataType::$cty(_) => $obj.extract::<$rty>().map(Value::$cty),
+            )*
+            _ => unreachable!(),
+        }
+    };
+    ($dtype:ident,$obj:ident, $($cty:ident =ord=> $rty:ty),*$(,)?) => {
+        match $dtype {
+            $(
+                ConcreteDataType::$cty(_) => $obj.extract::<$rty>()
+                .map(OrderedFloat)
+                .map(Value::$cty),
+            )*
+            _ => unreachable!(),
+        }
+    };
+}
+
+/// to int/float/boolean, if dtype is None, then convert to highest prec type
+pub(crate) fn pyo3_obj_try_to_typed_val(
+    obj: &PyAny,
+    dtype: Option<ConcreteDataType>,
+) -> PyResult<Value> {
+    if let Ok(b) = obj.downcast::<PyBool>() {
+        if let Some(ConcreteDataType::Boolean(_)) = dtype {
+            let dtype = ConcreteDataType::boolean_datatype();
+            let ret = to_con_type!(dtype, b,
+                Boolean => bool
+            )?;
+            Ok(ret)
+        } else {
+            Err(PyValueError::new_err(format!(
+                "Can't cast num to {dtype:?}"
+            )))
+        }
+    } else if let Ok(num) = obj.downcast::<PyInt>() {
+        if let Some(dtype) = dtype {
+            if dtype.is_signed() || dtype.is_unsigned() {
+                let ret = to_con_type!(dtype, num,
+                    Int8 => i8,
+                    Int16 => i16,
+                    Int32 => i32,
+                    Int64 => i64,
+                    UInt8 => u8,
+                    UInt16 => u16,
+                    UInt32 => u32,
+                    UInt64 => u64,
+                )?;
+                Ok(ret)
+            } else {
+                Err(PyValueError::new_err(format!(
+                    "Can't cast num to {dtype:?}"
+                )))
+            }
+        } else {
+            num.extract::<i64>().map(Value::Int64)
+        }
+    } else if let Ok(num) = obj.downcast::<PyFloat>() {
+        if let Some(dtype) = dtype {
+            if dtype.is_float() {
+                let ret = to_con_type!(dtype, num,
+                    Float32 =ord=> f32,
+                    Float64 =ord=> f64,
+                )?;
+                Ok(ret)
+            } else {
+                Err(PyValueError::new_err(format!(
+                    "Can't cast num to {dtype:?}"
+                )))
+            }
+        } else {
+            num.extract::<f64>()
+                .map(|v| Value::Float64(OrderedFloat(v)))
+        }
+    } else {
+        Err(PyValueError::new_err(format!(
+            "Can't cast {obj} to {dtype:?}"
+        )))
+    }
+}
+
+/// cast a columnar value into python object
+///
+/// | Rust   | Python          |
+/// | ------ | --------------- |
+/// | Array  | PyVector        |
+/// | Scalar | int/float/bool  |
```

Comment:
I think we should support string scalar right now.  It's important for TSDB, all the tags are the string type.